### ISSUE TITLE
Creating internal debugging tools?

### ISSUE THREAD

**Issue by farleyknight** [protoman/rockbot#23]
I've been thinking about debugging more and more, and particularly how other games that shipped to NES or SNES handled it. If I recall, it was common to find a "debug mode" by pressing a certain combination of buttons. Sonic I definitely remember had a secret code for accessing a debug mode.

In that mode, you could pretty much move where ever you wanted with the directional keys. This made it easy to debug something that happens at a particular stage.

I'm thinking we could so something similar? I was thinking that you could enter in a "secret" password that would unlock certain debugging features, like the ability to move anywhere you want on the page. Or maybe the ability to create new enemies.

What are your thoughts on this?

**Comment by farleyknight** [comment]
Here's a link describing the debug mode.

http://sonic.wikia.com/wiki/Debug_Mode

There's probably tons more if you search for Sonic and debug mode.

**Comment by protoman** [comment]
I remember playing Sonic 2 and using the debug.
I don't know how useful it would be, as we can change almost all aspects in
the game by using the editor, like adding npcs, ladders and such at a point
that is easy to test it.

On the other hand, as you said, moving the player to any stage or map would
indeed be nice. What do you think would be useful for testing?

## 

Iuri Fiedoruk, Software Developer

**Comment by farleyknight** [comment]
I think it would be useful for watching values as you're playing, instead of printing to a STDOUT. That way we don't have to comment out print-debug statements and instead update variables in some kind of HUD that is a layer on top of the current game.

For example, if an enemy is getting stuck in some weird spot (#9), I'd like to know what that enemy's internal values are (x, y, velocity, etc) looked like so I can more quickly see how to fix it. Same with animations, I'd like to know which animation frame Rockbot is currently in so I can debug the shoot & run issue you posted in #1.

Basically I'm seeing this as a way to more quickly diagnose bugs instead of using print-debug, since that can get pretty noisy after a while. That way I can test the game for a few hours, playing like I'm an ordinary user, and if I spot a bug, I can pop open the debugger HUD, and I can see what the issue might be right away.

I don't want to go crazy with this just yet. I think just being able to see inside important parts of the stateful stuff inside the player object and the monster object would be enough for now. Stage-level variables are a whole other matter.

**Comment by protoman** [comment]
Oh, that is a nice idea. Actually, when I first started coding Rockbot, I
was goind directly into Playstation 2, and at the time, I didn't had set up
a development environement for that (needs a ethernet cable and to boot up
pslink tool), so what I did was this:
https://www.youtube.com/watch?v=PVrFKcgwKwM

Each rectangle was a method that was executed without failing :)

## 

Iuri Fiedoruk, Software Developer

**Comment by farleyknight** [comment]
We have a quickly growing list of issues! I think we could probably talk about roadmap soon-ish, so we can get a sense of priorities and dependencies. I feel this HUD should probably go before some of the other issues here. If I could get a basic debug HUD going (just a layer on top of everything else with some text), then I can probably start adding values to it and slowly modifying the HUD as I encounter new bugs. We'd probably want to put the values at the very bottom or very top so it doesn't obscure the main game area.

**Comment by farleyknight** [comment]
Reading through the issues, it looks like #10, #21, #8, #2, #1, #9, #3, #7, and possibly #11 would benefit from a debug HUD. Others not so much, or it might apply but I'm not as familiar with the code.

**Comment by protoman** [comment]
I agree, it would be useful.
One problem is the low resolution to write values is not very good, we need
to test the ttf font to check how it shows with smaller sizes.

I'm currently working mistly on the editor and scenesShow lib right now,
but I can give it a better look in the weekend.

## 

Iuri Fiedoruk, Software Developer

**Comment by farleyknight** [comment]
So I tried some stuff out and I realized that it's not very useful unless you can show almost every piece of state in the app on one screen. 

Instead what I'm thinking about doing is adding another menu option to the config menu (next to audio, video, input), that's only available when a GAME_FLAGS bool has been turned on via `--debug-mode`. That menu could then drill down further and further into objects. Perhaps once I get that screen going we could have the ability to add stuff from the config menu onto the HUD permanently. Still experimenting with these ideas..
